app/utils/checkOrigin.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const {parse} = require('url');
+
+const {default: logger} = require('@financial-times/n-logger');
+
+const defaultMethods = [
+    'GET',
+    'POST',
+    'PUT',
+    'DELETE'
+];
+
+const defaultDomain = 'ft.com';
+
+module.exports = ({
+    methods = defaultMethods,
+    domain = defaultDomain
+} = {}) => (request, response, next) => {
+
+    const checkDomain = url =>
+        Boolean(url && parse(url).hostname.endsWith(`.${domain}`));
+
+    if (response.locals && response.locals.flags && response.locals.flags.disableSudsOriginCheck) {
+        return next();
+    }
+
+    const {
+        originalUrl,
+        method
+    } = request;
+
+    if (!methods.includes(method)) {
+        return next();
+    }
+
+    const origin = request.get('origin');
+    const referer = request.get('referer');
+
+    const originIsGood = checkDomain(origin);
+    const refererIsGood = checkDomain(referer);
+
+    if (!origin) {
+        logger.warn('Request does not have an origin.', {
+            event: 'NO_REQUEST_ORIGIN',
+            originalUrl,
+            method,
+            referer
+        });
+    } else if (!originIsGood) {
+        logger.warn(`Request origin's domain is not ${domain}.`, {
+            event: 'BAD_REQUEST_ORIGIN',
+            origin,
+            originalUrl,
+            method,
+            referer
+        });
+    }
+
+    if (!referer) {
+        logger.warn('Request does not have an referer.', {
+            event: 'NO_REQUEST_REFERER',
+            origin,
+            originalUrl,
+            method
+        });
+    } else if (!refererIsGood) {
+        logger.warn(`Request referer's domain is not ${domain}.`, {
+            event: 'BAD_REQUEST_REFERER',
+            origin,
+            originalUrl,
+            method,
+            referer
+        });
+    }
+
+
+    if (originIsGood && refererIsGood) {
```

Comment:
I think if we convert this to...
```suggestion
    if (originIsGood || refererIsGood) {
```
should automatically cover the next 2 if statements, what do you think?

Reply:
If you consider this example where the origin is bad but the referer is good, this suggestion would allow that access. We only want it to allow access if the referer is good **and** the origin is good or undefined

Reply:
It it not logically the same thing. For example, if the referrer is provided but it doesn't match the domain (e.g because of some hacking attempt), we want to block that, even if the origin is good.